examples/dao/assets/dao-app-approval.py

```diff
@@ -0,0 +1,471 @@
+        Assert(Global.creator_address() == Txn.sender()),
+        App.globalPut(deposit_lsig, Txn.application_args[1]),
+        Return(Int(1))
+    ])
+
+    # A proposal is submitted using an lsig and recorded into that lsig using add_proposal.
+    # Expected arguments: proposal_config
+    # [url, url_hash, voting_start, voting_end, execute_before, type,
+    # ALGO/ASA transfer params or message (according to type)]
+    add_proposal = Seq([
+        # When recording a proposal, we fail if there is a proposal recorded in the account
+        # User should call close_proposal use-case to remove a not active proposal record and withdraw deposit.
+        Assert(App.localGet(Int(0), Bytes("type")) == Int(0)),
+        # Verify deposit to deposit_lsig equals global.deposit
+        verify_deposit(Int(1), App.globalGet(deposit_lsig)),
+        Assert(Gtxn[1].asset_amount() == App.globalGet(deposit)),
+
+        # if everything goes well, start recording proposal info
+        App.localPut(Int(0), Bytes("name"), Txn.application_args[1]),
+        App.localPut(Int(0), Bytes("url"), Txn.application_args[2]),
+        App.localPut(Int(0), Bytes("url_hash"), Txn.application_args[3]),
+        If(
+            Txn.application_args[4] == Bytes(""),
+            App.localPut(Int(0), Bytes("hash_algo"), Bytes("sha256")), # default hash_algo
+            App.localPut(Int(0), Bytes("hash_algo"), Txn.application_args[4])
+        ),
+        # save some config in scratch first (to reuse later)
+        scratchvar_voting_start.store(Btoi(Txn.application_args[5])),
+        scratchvar_voting_end.store(Btoi(Txn.application_args[6])),
+        scratchvar_execute_before.store(Btoi(Txn.application_args[7])),
+        scratchvar_proposal_type.store(Btoi(Txn.application_args[8])),
+
+        # voting_start must be after now
+        Assert(scratchvar_voting_start.load() > Global.latest_timestamp()),
+        App.localPut(Int(0), Bytes("voting_start"), scratchvar_voting_start.load()),
+        Assert(And(
+            # voting_end must be > voting_start
+            scratchvar_voting_end.load() > scratchvar_voting_start.load(),
+            # min_duration <= voting_end - voting_start <= max_duration
+            App.globalGet(min_duration) <= scratchvar_voting_end.load() - scratchvar_voting_start.load(),
+            App.globalGet(max_duration) >= scratchvar_voting_end.load() - scratchvar_voting_start.load()
+        )),
+        App.localPut(Int(0), Bytes("voting_end"), scratchvar_voting_end.load()),
+        # execute_before must be after voting_end
+        Assert(scratchvar_execute_before.load() > scratchvar_voting_end.load()),
+        App.localPut(Int(0), Bytes("execute_before"), scratchvar_execute_before.load()),
+        # type must be 1, 2 OR 3
+        Assert(
+            Or(
+                scratchvar_proposal_type.load() == Int(1),
+                scratchvar_proposal_type.load() == Int(2),
+                scratchvar_proposal_type.load() == Int(3)
```

Comment:
do you know if this is an optimal way? loading 3 times doesn't create lot of TEAL code? Ideally we should load once and check top of the stack against 3 values.

Reply:
`scratch.load` loads a [copy](https://developer.algorand.org/docs/reference/teal/opcodes/#load-i) from scratch slot X to top of stack (cost of both `load X` & `store X` is 1). The other way would be to load once, create 2 duplicates (using `dup`), and then compare.
Current:
```teal
load X // loads a copy from scratch slot X
int 1
==
load X
int 2
==
||
load X
int 3
==
||
```
Using `dup`:
```
load X
dup
dup
int 1
==
int 2 
== 
||
int 3
==
||
```
both have same cost & length. Will update if you prefer the latter. Let me know

Reply:
OK, the first one is simpler to read I think. So you can leave.